src/target/adi.c

```diff
@@ -586,14 +591,147 @@ static void adi_parse_adi_rom_table(adiv5_access_port_s *const ap, const target_
 		if (entry == 0)
 			break;
 
-		if (!(entry & ADIV5_ROM_ROMENTRY_PRESENT)) {
+		if (!(entry & ADIV5_ROM_CLASS1_ROMENTRY_PRESENT)) {
 			DEBUG_INFO("%s%" PRIu32 " Entry 0x%08" PRIx32 " -> Not present\n", indent, i, entry);
 			continue;
 		}
 
 		/* Probe recursively */
-		adi_ap_component_probe(ap, base_address + (entry & ADIV5_ROM_ROMENTRY_OFFSET), recursion_depth + 1U, i);
+		adi_ap_component_probe(ap, base_address + (entry & ADIV5_ROM_CLASS1_ROMENTRY_OFFSET), recursion_depth + 1U, i);
+	}
+	DEBUG_INFO("%sROM Table: END\n", indent);
+}
+
+static bool adi_reset_resources(adiv5_access_port_s *const ap, const target_addr64_t base_address)
+{
+	/* Read out power request ID register 0 and check if power control is actually implemented */
+	const uint8_t pridr0 = adi_mem_read32(ap, base_address + ADIV5_ROM_CLASS9_PRIDR0) & 0x3fU;
+	if ((pridr0 & ADIV5_ROM_CLASS9_PRIDR0_VERSION_MASK) != ADIV5_ROM_CLASS9_PRIDR0_VERSION_NOT_IMPL)
+		ap->flags |= ADIV6_DP_FLAGS_HAS_PWRCTRL;
+	/* Now try and perform a debug reset request */
+	if (pridr0 & ADIV5_ROM_CLASS9_PRIDR0_HAS_DBG_RESET_REQ) {
+		platform_timeout_s timeout;
+		platform_timeout_set(&timeout, 250);
+
+		adi_mem_write32(ap, base_address + ADIV5_ROM_CLASS9_DBGRSTRR, ADIV5_ROM_CLASS9_DBGRST_REQ);
+		/* While the reset request is in progress */
+		while (adi_mem_read32(ap, base_address + ADIV5_ROM_CLASS9_DBGRSTRR) & ADIV5_ROM_CLASS9_DBGRST_REQ) {
+			/* Check if it's been acknowledge, and if it has, deassert the request */
+			if (adi_mem_read32(ap, base_address + ADIV5_ROM_CLASS9_DBGRSTAR) & ADIV5_ROM_CLASS9_DBGRST_REQ)
+				adi_mem_write32(ap, base_address + ADIV5_ROM_CLASS9_DBGRSTRR, 0U);
+			/* Check if the reset has timed out */
+			if (platform_timeout_is_expired(&timeout)) {
+				DEBUG_WARN("adi: debug reset failed\n");
+				adi_mem_write32(ap, base_address + ADIV5_ROM_CLASS9_DBGRSTRR, 0U);
+				break;
+			}
+		}
+	}
+	/* Regardless of what happened, extract whether system reset is supported this way */
+	if (pridr0 & ADIV5_ROM_CLASS9_PRIDR0_HAS_SYS_RESET_REQ)
+		ap->flags |= ADIV6_DP_FLAGS_HAS_SYSRESETREQ;
+	return true;
+}
+
+static inline uint64_t adi_read_rom_class_9_entry(
+	adiv5_access_port_s *const ap, const uint8_t rom_format, const target_addr64_t entry_address)
+{
+	const uint32_t entry_lower = adi_mem_read32(ap, entry_address);
+	if (rom_format == ADIV5_ROM_CLASS9_DEVID_FORMAT_32BIT)
+		return entry_lower;
+	const uint32_t entry_upper = adi_mem_read32(ap, entry_address + 4U);
+	return ((uint64_t)entry_upper << 32U) | (uint64_t)entry_lower;
+}
+
+static void adi_parse_adi_rom_table_class9(adiv5_access_port_s *const ap, const target_addr64_t base_address,
```

Comment:
This isn't an ADI ROM table, but a CoreSight ROM table - please use the nomenclature `adi_parse_coresight_rom_table` (this also removes the need to rename the other parser function above).

While the ADIv6 spec defines how these tables work, the documentation in question is lifted from the CoreSight specification and TRMs, hence why this particular choice in nomenclature.